File: Strings/1832_Check_If_The_Sentence_Is_Pangram.java
/*
1832. Check if the Sentence Is Pangram

A pangram is a sentence where every letter of the English alphabet appears at least once.
Given a string sentence containing only lowercase English letters, return true if sentence is a pangram, or false otherwise.

Example 1: Input: sentence = "thequickbrownfoxjumpsoverthelazydog" Output: true
Explanation: sentence contains at least one of every letter of the English alphabet.
Example 2: Input: sentence = "leetcode" Output: false

Difficulty : easy

Approach : Array.
1. Create an integer array of size 26 to keep track of the frequency of each letter in the sentence.
2. Iterate through each character in the sentence, and for each character, calculate its index (
0-25) by subtracting 'a' from the character. Increment the corresponding index in the frequency array.
3. After processing all characters, check if all indices in the frequency array are greater than
    zero. If any index is zero, it means that letter is missing from the sentence, and we return false. Otherwise, return true.

Time Complexity: O(n) where n is the length of the sentence, since we iterate through each character once.
Space Complexity: O(1) since the size of the frequency array is constant (26).

*/

class Solution {
    public boolean checkIfPangram(String sentence) {
        
        // boolean[] seen = new boolean[26];
        // for (int i = 0; i < sentence.length(); i++) {
        //     char ch = sentence.charAt(i);
        //     seen[ch - 'a'] = true;
        // }
        // for (int i = 0; i < 26; i++) {
        //     if (!seen[i]){
        //         return false;
        //     }
        // }

        // return true;
        int[] freq = new int[26];

        for(int i = 0; i < sentence.length(); i++){
            int index = sentence.charAt(i) - 'a';
            freq[index]++;
        }

        for(int i = 0; i < 26; i++){
            if(freq[i] == 0){
                return false;
            }
        }

        return true;
    }
}

File: Strings/1832_Check_If_The_Sentence_Is_Pangram.py
"""
1832. Check if the Sentence Is Pangram

A pangram is a sentence where every letter of the English alphabet appears at least once.
Given a string sentence containing only lowercase English letters, return true if sentence is a pangram, or false otherwise.

Example 1: Input: sentence = "thequickbrownfoxjumpsoverthelazydog" Output: true
Explanation: sentence contains at least one of every letter of the English alphabet.
Example 2: Input: sentence = "leetcode" Output: false

Difficulty : easy

Approach :
1. Create a set to store the unique characters in the sentence.
2. Iterate through each character in the sentence and add it to the set.
3. After processing all characters, check the size of the set. If it contains 26 unique characters, return true; otherwise, return false.
4. This approach has a time complexity of O(n), where n is the length of the sentence, and a space complexity of O(1) since the set can contain at most 26 characters.
5. The implementation of this approach is straightforward and can be done in a few lines of code.
The code below implements this approach in Python.
Time Complexity : O(n)
Space Complexity : O(1)

"""

class Solution:
    def checkIfPangram(self, sentence: str) -> bool:
        seen = set()

        for ch in sentence:
            seen.add(ch)

        return len(seen) == 26

        # freq = [0] * 26

        # for ch in sentence:
        #     index = ord(ch) - ord('a')
        #     freq[index] += 1

        # for i in range(26):
        #     if freq[i] == 0:
        #         return False

        # return True